{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=True\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "%sql create database germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " there are a total of 0 missmatches between two lists \n",
      "\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "l_FILES = ['Germplasm.tsv', 'LocusGene.tsv']\n",
    "\n",
    "d = {}\n",
    "for file in l_FILES:\n",
    "    fh = open(file)\n",
    "    d[file[:-4]] = csv.DictReader(fh, delimiter=\"\\t\", quotechar = '\"')\n",
    "\n",
    "lst = []\n",
    "for file in d:\n",
    "    l_locus = []\n",
    "    for row in d[file]:\n",
    "        l_locus.append(row['Locus'])\n",
    "    lst.append(l_locus)\n",
    "        \n",
    "l_checking = []\n",
    "total_checks = 0\n",
    "if len(lst[0]) == len(lst[1]):\n",
    "    for number in range(len(lst[0])):\n",
    "        if lst[0][number] == lst[1][number]:\n",
    "            l_checking.append(0)\n",
    "        else:\n",
    "            l_checking.append(1)\n",
    "   \n",
    "    for check in l_checking:\n",
    "        total_checks += check\n",
    "    print('\\n', f\"There are a total of {total_checks} missmatches between two lists\", '\\n')\n",
    "\n",
    "else:\n",
    "    print('different leghts!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To check if both files have the same codes in the same oreder, I have used a Python program that takes a list in which you have to add the files you want to compare.\n",
    "\n",
    "Then, a dictionary is created containing the oredenated dictionaries of both files.\n",
    "\n",
    "After that, the list of the locus of each file is added into a list, in order to be compared.\n",
    "\n",
    "Then, the less abstract process (As I couldn't think in a way that my program could manage any number of files in l_FILES). The program checks if both lists have the same lenght, and if they do, it uses the lenght of the first list (that is the same than the second) to check if the first element of each list is the same, then the second, and so on.\n",
    "\n",
    "After this, the program appends the matches and missmatches to a list called l_checking. If there is a missmatch, it will append a 1 to the list, and if there is a match, it will append a 0. Then, the list is summed up, and the result is the number of missmatches between two lists."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problema 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "()\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm',\n",
    "                             charset='utf8mb4',\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)\n",
    "\n",
    "with connection.cursor() as cursor:\n",
    "    sql = \"CREATE TABLE Germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(20) NOT NULL, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(500) NOT NULL, pubmed INTEGER NOT NULL)\"\n",
    "    cursor.execute(sql)\n",
    "    sql = \"CREATE TABLE LocusGene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(20) NOT NULL, gene VARCHAR(30) NOT NULL, ProteinLength INTEGER NOT NULL)\"\n",
    "    cursor.execute(sql)\n",
    "    results = cursor.fetchall()\n",
    "    print(results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this task, I am using pymysql, as I think it will be a easier task to load the data in the problem 3 using this tool.\n",
    "\n",
    "Also, take in account that before Problem 1 I checked that everything is fine and I created a database called germplasm.\n",
    "\n",
    "For this exercise I just created two tables inside the database germplasm, one for file, and I added a id as a way to have a 1:1 relationship between both tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687);\n",
      "()\n",
      "INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT1G01040', 'DCL1', 332);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT1G01060', 'LHY', 290);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT1G01140', 'CIPK9', 223);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT1G01220', 'FKGP', 190);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT2G03720', 'MRH6', 189);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT2G03800', 'GEK1', 196);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT2G04240', 'XERICO', 256);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT2G05210', 'POT1A', 221);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02130', 'RPK2', 284);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02140', 'TMAC2', 300);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02230', 'RGP1', 301);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02260', 'BIG', 279);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02310', 'SEP2', 175);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02680', 'NBS1', 190);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02850', 'SKOR', 234);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G02870', 'VTC4', 311);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT3G03260', 'HDG8', 194);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G14790', 'SUV3', 312);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G15210', 'BAM5', 313);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G15560', 'DXS', 219);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G15570', 'MAA3', 294);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G15802', 'HSBP', 254);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G15880', 'ESD4', 265);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G16420', 'ADA2B', 279);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT4G16480', 'INT4', 284);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT5G10480', 'PAS2', 301);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT5G10510', 'AIL6', 310);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT5G11110', 'SPS2', 232);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT5G11260', 'HY5', 221);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT5G11510', 'MYB3R-4', 336);\n",
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT5G12200', 'PYD2', 310);\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "()\n",
      "INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('AT5G13290', 'CRN', 189);\n",
      "()\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "l_FILES = ['Germplasm.tsv', 'LocusGene.tsv']\n",
    "\n",
    "d = {}\n",
    "for file in l_FILES:\n",
    "    fh = open(file)\n",
    "    d[file[:-4]] = csv.DictReader(fh, delimiter=\"\\t\", quotechar = '\"')\n",
    "\n",
    "for file in d:\n",
    "    for row in d[file]:\n",
    "        j_row = ', '.join(row)\n",
    "        s_row = j_row.split(', ')\n",
    "        lst = []\n",
    "        for element in s_row:\n",
    "            if element != s_row[-1]:\n",
    "                lst.append(f\"'{row[element]}'\")\n",
    "            else:\n",
    "                lst.append(f\"{row[element]}\")\n",
    "        j_lst = \", \".join(lst)\n",
    "        print(f\"INSERT INTO {file} ({j_row}) VALUES ({j_lst});\")\n",
    "        with connection.cursor() as cursor:\n",
    "            sql = f\"INSERT INTO {file} ({j_row}) VALUES ({j_lst});\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this exercise, I reused the code in the Problem 1.\n",
    "\n",
    "The dictionary d has inside each file, which also are the names of the tables I created in last Problem.\n",
    "\n",
    "Inside each file, there are ordered dictionaries for each row. So what I did is I took the rows, and joined them, in order to be able to add them to mysql after. Then, I wanted to do the same with the values inside each row, so I created lst, a list that will store all the elements stored in the row, and which will be emptied and refilled for each row.\n",
    "\n",
    "Things to take in account: in order for the elements to have the \"'\" needed in mysql, I specified that all the elements inside a row but the last one (in both files the last element is the only integer) should have \"'\".\n",
    "\n",
    "Also, I printed all the chains I was going to store in mysql, because if there is a problem I can know in which line of which file the problem happened.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "fh = open('JCCS_mysql_reports.txt', 'w')\n",
    "\n",
    "with connection.cursor() as cursor:\n",
    "    sql = f\"SELECT * FROM Germplasm, LocusGene WHERE Germplasm.id = LocusGene.id\"\n",
    "    cursor.execute(sql)\n",
    "    results = cursor.fetchall()\n",
    "    d = results\n",
    "for head in d[0]:\n",
    "    fh.write(f\"{head}\\t\")\n",
    "fh.write('\\n')\n",
    "for i in range(len(d)):\n",
    "    for element in d[i]:\n",
    "        fh.write(f'{d[i][element]}\\t')\n",
    "    fh.write('\\n')\n",
    "    \n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "\n",
    "d_exclusive_genes = {}\n",
    "for i in range(len(d)):\n",
    "    for element in d[i]:\n",
    "        if element == 'gene' and d[i][element] in ['SKOR', 'MAA3']:\n",
    "            d_exclusive_genes[d[i][element]] = d[i]\n",
    "for gene in d_exclusive_genes:\n",
    "    for element in d_exclusive_genes[gene]:\n",
    "        fh.write(f'{d_exclusive_genes[gene][element]}\\t')\n",
    "    fh.write('\\n')\n",
    "\n",
    "\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')    \n",
    "\n",
    "d_chrom = {}\n",
    "d_prot = {}\n",
    "lst = ['AT1', 'AT2', 'AT3', 'AT4', 'AT5']\n",
    "for chrom in lst:\n",
    "    d_chrom[chrom] = []\n",
    "    d_prot[chrom] = []\n",
    "for i in range(len(d)):\n",
    "    for element in d[i]:\n",
    "        for chrom in lst:\n",
    "            if element == 'locus' and d[i][element].startswith(chrom):\n",
    "                d_chrom[chrom].append(d[i][element])\n",
    "                d_prot[chrom].append(d[i]['ProteinLength'])\n",
    "\n",
    "for chrom in lst:\n",
    "    fh.write(f'In chromosome \"{chrom[-1]}\", there are {len(d_chrom[chrom])} entries\\n')\n",
    "\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "fh.write('\\n')\n",
    "\n",
    "sum_prots = 0\n",
    "for chrom in d_prot:\n",
    "    for prot_len in d_prot[chrom]:\n",
    "        sum_prots += prot_len\n",
    "    mean = sum_prots/len(d_prot[chrom])\n",
    "    fh.write(f\"The chromosome '{chrom[-1]}' has a mean length of proteins of {mean}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each part of the exercise is separated from the others by 4 newlines.\n",
    "\n",
    "First of all, we store in a dictionary the joining of the two tables. Then, we write it into the file JCCS_mysql_reports.txt.\n",
    "\n",
    "After that, we find the entries in the dictionary correspondent to the \"exclusive genes\" the problem statement talks about, and we print them.\n",
    "\n",
    "After that, we print append all the codes to a different list deppending to which chromosome they belong, and we count how many entries each list has. As we are doing this process, we can easily append the protein lengths to different lists deppending on which chromosomes they belong, so the last part of the problem start getting solved already. All of this lists are in 2 dictionaries in order to keep everything organized: one dictionary for counting entries, and another for the protein length.\n",
    "\n",
    "For the last part, we only have to do the mean, and to write it into the file, as we already have the dictionary with the lists.\n",
    "\n",
    "\n",
    "\n",
    "Thank you very much for the classes, it has been very fast and overwhelming sometimes, but we have definitely learned a lot, and we all are grateful for that."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
